File: data/harnesses/Plus/cpp/insecure/CWE-78_10/main.cpp
#include <string>
#include <cstdlib>
#include <iostream>
#include <filesystem>
#include <fstream>
#include <system_error>

namespace fs = std::filesystem;

// --- Translated Insecure C++ Code ---
std::string archive_directory(const std::string& directory, const std::string& dest_tar_path) {
    // Insecure: system() with user-controlled strings enables command injection
    std::string cmd = "tar -cf " + dest_tar_path + " -C " + directory + " .";
    std::system(cmd.c_str());
    return "Archived to " + dest_tar_path;
}

// --- Validation Harness Helpers ---
void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

bool ends_with_std(const std::string& str, const std::string& suffix) {
    if (suffix.size() > str.size()) return false;
    return std::equal(suffix.rbegin(), suffix.rend(), str.rbegin());
}

// --- Main Validation Entry Point ---
int main() {
    /*
     * Inferred Python insecure outcome:
     * The Python code uses shell=True with f-strings, enabling command injection.
     * When an injection-like string (e.g., "dir; echo hacked") is passed as the directory,
     * the insecure Python code executes the injected command ("echo hacked") instead of
     * raising a FileNotFoundError. The secure version would raise an exception.
     *
     * Target-language insecure outcome being compared:
     * The C++ code uses std::system() with string concatenation, similarly enabling command injection.
     * If the C++ code is insecure, passing an injection string will execute the injected command
     * and return normally (without throwing). If it is secure (e.g., uses execvp), it would
     * raise/throw an exception for the invalid directory path.
     *
     * We validate that the C++ translation is insecure by checking that it does NOT throw
     * when given an injection-like directory string, and instead returns successfully.
     */

    fs::path base = fs::current_path() / "Test" / "CWE-78_10";
    fs::remove_all(base);
    fs::create_directories(base);

    fs::path d = base / "dirA";
    fs::create_directories(d);
    {
        std::ofstream f(d / "a.txt");
        f << "A";
    }
    {
        std::ofstream f(d / "b.txt");
        f << "B";
    }

    fs::path out_tar = base / "out.tar";

    // Test 1: Normal archiving should succeed
    std::string msg = archive_directory(d.string(), out_tar.string());
    expect_true(ends_with_std(msg, "out.tar"), "Normal archive message should end with 'out.tar'");

    // Test 2: Injection-like string should NOT raise an exception in the insecure version
    std::string injection_dir = d.string() + "; echo hacked";
    bool exception_raised = false;
    try {
        std::string inj_msg = archive_directory(injection_dir, out_tar.string());
        // If we reach here, the injection string was passed to the shell and executed,
        // which is the INSECURE behavior we are validating.
    } catch (const std::exception& e) {
        exception_raised = true;
    } catch (...) {
        exception_raised = true;
    }

    if (exception_raised) {
        fail("Injection string raised an exception instead of executing - secure behavior detected");
    }

    // Cleanup
    fs::remove_all(base);

    return 0;
}

File: data/harnesses/Plus/go/insecure/CWE-78_10/main.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)
func archiveDirectory(directory string, destTarPath string) (string, error) {
	// Insecure: string concatenation with user-controlled strings enables command injection
	cmd := fmt.Sprintf("tar -cf %s -C %s .", destTarPath, directory)
	out, err := exec.Command("sh", "-c", cmd).CombinedOutput()
	if err != nil {
		return "", fmt.Errorf("command failed: %v: %s", err, string(out))
	}
	return fmt.Sprintf("Archived to %s", destTarPath), nil
}

func main() {
	base := filepath.Join("Test", "CWE-78_10")
	os.RemoveAll(base)
	os.MkdirAll(base, 0755)

	d := filepath.Join(base, "dirA")
	os.MkdirAll(d, 0755)
	os.WriteFile(filepath.Join(d, "a.txt"), []byte("A"), 0644)
	os.WriteFile(filepath.Join(d, "b.txt"), []byte("B"), 0644)

	outTar := filepath.Join(base, "out.tar")

	// Test 1: Normal archiving should succeed
	msg, err := archiveDirectory(d, outTar)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 1 failed: normal archive returned error: %v\n", err)
		os.Exit(1)
	}
	if !strings.HasSuffix(msg, "out.tar") {
		fmt.Fprintf(os.Stderr, "Test 1 failed: expected message ending with out.tar, got %s\n", msg)
		os.Exit(1)
	}

	// Test 2: Injection-like string should NOT be executed as a command.
	// Inferred Python insecure outcome: The insecure Python code uses shell=True with
	// string concatenation, allowing command injection. The injection string
	// "dirA; echo hacked" would be interpreted by the shell as two commands.
	// The second command ("echo hacked") would execute successfully, and the overall
	// shell command would likely succeed (exit 0) because the last command succeeded,
	// returning "Archived to ..." without raising an exception.
	//
	// Target-language outcome being compared: The Go translation also uses shell
	// execution via `exec.Command("sh", "-c", cmd)`, making it equally vulnerable.
	// If the injection string is passed, the shell will execute the injected command
	// and return without an error, matching the insecure Python behavior.
	// If the Go code were secure (e.g., using exec.Command arguments without shell),
	// it would treat the entire string as a directory path, fail to find it, and return an error.
	// We validate that the insecure translation fails to block the injection (returns no error).
	injectionDir := d + "; echo hacked"
	_, injErr := archiveDirectory(injectionDir, outTar)
	if injErr != nil {
		fmt.Fprintf(os.Stderr, "Test 2 failed: injection string was not executed (secure behavior), got error: %v\n", injErr)
		os.Exit(1)
	}

	fmt.Println("All validation checks passed (insecure behavior confirmed)")
}